lib/Dialect/Torch/Transforms/ReduceOpVariants.cpp

```diff
@@ -59,6 +61,10 @@ class ConvertToImmutableTensors : public RewritePattern {
             listConstruct.elements(), [&](Value tensor) -> Value {
               return rewriter.create<CopyToValueTensorOp>(op->getLoc(), tensor);
             }));
+        // Checking if the OptionalType is a NonValueTensorType optional type.
+        if (listType.getContainedType().isa<OptionalType>() &&
+            newListElements.front().getType().isa<NonValueTensorType>())
```

Comment:
I don't think this is right. We should be checking all the elements in the list. This should also happen before we perform the `CopyToValueTensorOp` because by the time we copy the tensor we should be sure that it is not none.  I think we should do what @cathyzhyi mentioned, which is to handle each optional tensor the same way that it is done here

https://github.com/llvm/torch-mlir/blob/28be3d84ce0735c45a215e88729ae63e5ac7f1d9/lib/Dialect/Torch/Transforms/ReduceOpVariants.cpp#L72

All the checks happen before we `CopyToValueTensor`. Note: This will more than likely result in a lot of shared code, so I think it would be good to have a helper function that takes care of checking that an optional tensor is valid for copying to value tensor and use that here and in the `else-if` branch below that handles optional types.

Also, it would be nice to have some sort of test to make sure things are indeed working as expected. You can add a test like this one

https://github.com/llvm/torch-mlir/blob/4486de5ef3f31e2ae45b1a557823b4f635694631/test/Dialect/Torch/reduce-op-variants.mlir#L70